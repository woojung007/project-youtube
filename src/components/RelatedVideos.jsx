import React from 'react';
import { useYoutubeApi } from 'context/YoutubeApiContext';
import { useQuery } from '@tanstack/react-query';
import VideoCard from 'components/VideoCard';

export default function RelatedVideos({ id }) {
    const { youtube } = useYoutubeApi();
    const { error, isLoading, data: videos } = useQuery(['related', id], () => youtube.relatedVideos(id));

    return (
        <>
            {isLoading && <p>Loading...</p>}
            {error && <p>Something is wrong ðŸ˜…</p>}
            {videos && (
                <ul className=''>
                    {videos.map((video) => (
                        <VideoCard type='list' key={video.id} video={video} />
                    ))}
                </ul>
            )}
        </>
    );
}
