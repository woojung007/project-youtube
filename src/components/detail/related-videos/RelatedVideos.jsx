import React from 'react';

export default function RelatedVideos() {
    return (
        <div>
            <div>related videos</div>
        </div>
    );
}
