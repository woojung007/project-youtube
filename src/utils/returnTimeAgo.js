import moment from '../../node_modules/moment/moment';

export const returnTimeAgo = (date) => {
    const now = new Date();

    return moment(date).from(moment(now));
};
