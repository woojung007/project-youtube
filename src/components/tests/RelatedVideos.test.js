/* eslint-disable testing-library/prefer-find-by */
import '@testing-library/jest-dom';
import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import RelatedVideos from 'components/RelatedVideos';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from 'tests/utils';
import { fakeVideos } from 'tests/videos';

describe('RelatedVideos', () => {
    const fakeYoutube = {
        relatedVideos: jest.fn(),
    };

    afterEach(() => {
        fakeYoutube.relatedVideos.mockReset();
    });

    it('renders correctly', async () => {
        fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
        const { asFragment } = renderRelatedVideos();

        await waitForElementToBeRemoved(screen.queryByText('Loading...'));
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders related videos correctly', async () => {
        fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
        renderRelatedVideos();

        expect(fakeYoutube.relatedVideos).toHaveBeenCalledWith('videoId');
        await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(fakeVideos.length));
    });

    it('renders loading', () => {
        fakeYoutube.relatedVideos.mockImplementation(() => fakeVideos);
        renderRelatedVideos();

        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('renders error', async () => {
        fakeYoutube.relatedVideos.mockImplementation(() => {
            throw new Error('error');
        });

        renderRelatedVideos();

        await waitFor(() => expect(screen.getByText('Something is wrong ðŸ˜…')).toBeInTheDocument());
    });

    function renderRelatedVideos() {
        return render(
            withAllContexts(withRouter(<Route path='/' element={<RelatedVideos id='videoId' />} />), fakeYoutube),
        );
    }
});
