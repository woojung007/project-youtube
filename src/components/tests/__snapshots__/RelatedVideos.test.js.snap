// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideos renders correctly 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title"
        class="w-60 mr-2"
        src="http://image/"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
        >
          title
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
    <li
      class="flex gap-1 m-2"
    >
      <img
        alt="title2"
        class="w-60 mr-2"
        src="http://image2/"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle2
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
